### ISSUE TITLE
Holistic cross-correlation matrix

### ISSUE THREAD

**Issue by AChatzigoulas** [sukritsingh/cardsReader#1]
Dear enspara Team,

Is there a fast and efficient way to produce the holistic cross-correlation matrix? I did ti by running target_site_MI for every single residue, but it takes a long time for a large protein.

Thank you very much in advance. # #

**Comment by sukritsingh** [comment]
Huge apologies for the delayed response! I somehow missed this across a move and few other personal-life events....

>Is there a fast and efficient way to produce the holistic cross-correlation matrix?  

Yes! It's actually as simple as summing up the four matrices output by `collect_cards.py` in enspara! The output from enspara is a dict containing the four matrices (and their labels). Summing all 4 would generate the holistic cross-correlation matrix (and summing specific pairs will provide other useful matrices as well). 

`target_site_MI` is a useful metric for computing communication from that summed matrix, but it runs by iterating across every residue in the protein for a given target site. One approach here might be to utilize a multiprocessing wrapper that runs each residue as a target site on a single core? 

Thanks!

**Comment by AChatzigoulas** [comment]
Thank you very much for the response. Yes I guess a multiprocess wrapper is probably the easiest solution. 

Cheers

**Comment by sukritsingh** [comment]
Of course! 

Keep in mind that target_site_MI is also driven with the desire to find a known target site. It sounds like you're looking for more global communication structures within your protein? In that case, one thing we've done often and had success with is using[ Affinity Propagation](https://scikit-learn.org/stable/modules/generated/sklearn.cluster.AffinityPropagation.html) to cluster the residue-level MI matrix (so sum the holistic MI matrix up based on which dihedrals are all in the same residue). We've had success with that identifying correlated motions in [small viral proteins](https://www.biorxiv.org/content/10.1101/2020.02.09.940510v1.abstract) and[ 300-residue sized enzymes](https://www.biorxiv.org/content/10.1101/2021.04.14.439842v1). I have some code to do Affinity Propagation on CARDS data that I'll be adding to the repo soon.

Hope this helps!

**Comment by AChatzigoulas** [comment]
I am more interested in finding allosteric networks. Something like the classic [Dynamical Network Analysis](https://doi.org/10.1073/pnas.0810961106), but with CARDS instead of Pearson correlations. So, I need a ```n_residues*n_residues``` matrix with the residue pairwise CARDS.

**Comment by sukritsingh** [comment]
This is exactly what CARDS was made for! The raw output of the Holistic correlation matrix is the dihedral pairwise CARDS values, but you can easily sum up by dihedrals to make the residue-level matrix, or use Target Site MI to generate this matrix (if you want to account for neighbor effects!). 

This residue-pairwise CARDS matrix can also be thought of as an adjacency matrix for a graph, in which case clustering algorithms like Affinity Propagation will allow you to identify groups of residues that act as "communities", similarly to dynamical network analysis! 

Hope this helps!

**Comment by AChatzigoulas** [comment]
Yes, this is very helpful indeed. Thank you very much. It would be awesome to automate this whole process, including clustering and further analysis (plots etc). Maybe a jupyter notebook with the code you used in your papers, if that's ok with you?
